src/external_trait_impls/rayon/raw.rs

```diff
@@ -0,0 +1,184 @@
+use alloc::alloc::{dealloc, Layout};
+use core::mem;
+use core::ptr::NonNull;
+use raw::Bucket;
+use raw::{RawIterRange, RawTable};
+use rayon::iter::{
+    plumbing::{self, Folder, UnindexedConsumer, UnindexedProducer},
+    ParallelIterator,
+};
+use scopeguard::guard;
+
+/// Parallel iterator which returns a raw pointer to every full bucket in the table.
+pub struct RawParIter<T> {
+    iter: RawIterRange<T>,
+}
+
+unsafe impl<T> Send for RawParIter<T> {}
+
+impl<T> ParallelIterator for RawParIter<T> {
+    type Item = Bucket<T>;
+
+    #[inline]
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        let producer = ParIterProducer { iter: self.iter };
+        plumbing::bridge_unindexed(producer, consumer)
+    }
+}
+
+/// Producer which returns a `Bucket<T>` for every element.
+struct ParIterProducer<T> {
+    iter: RawIterRange<T>,
+}
+
+unsafe impl<T> Send for ParIterProducer<T> {}
+
+impl<T> UnindexedProducer for ParIterProducer<T> {
+    type Item = Bucket<T>;
+
+    #[inline]
+    fn split(self) -> (Self, Option<Self>) {
+        let (left, right) = self.iter.split();
+        let left = ParIterProducer { iter: left };
+        let right = right.map(|right| ParIterProducer { iter: right });
+        (left, right)
+    }
+
+    #[inline]
+    fn fold_with<F>(self, folder: F) -> F
+    where
+        F: Folder<Self::Item>,
+    {
+        folder.consume_iter(self.iter)
+    }
+}
+
+/// Parallel iterator which consumes a table and returns elements.
+pub struct RawIntoParIter<T> {
+    iter: RawIterRange<T>,
+    alloc: Option<(NonNull<u8>, Layout)>,
+}
+
+unsafe impl<T> Send for RawIntoParIter<T> {}
+
+impl<T: Send> ParallelIterator for RawIntoParIter<T> {
+    type Item = T;
+
+    #[inline]
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        let _guard = guard(self.alloc, |alloc| {
+            if let Some((ptr, layout)) = *alloc {
+                unsafe {
+                    dealloc(ptr.as_ptr(), layout);
+                }
+            }
+        });
+        let producer = ParDrainProducer { iter: self.iter };
+        plumbing::bridge_unindexed(producer, consumer)
+    }
+}
+
+/// Parallel iterator which consumes elements without freeing the table storage.
+pub struct RawParDrain<'a, T> {
+    table: &'a mut RawTable<T>,
+}
```

Comment:
If this struct gets dropped without ever driving it, then it won't have drained any items.